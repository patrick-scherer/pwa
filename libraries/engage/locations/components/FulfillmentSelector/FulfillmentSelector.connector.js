// @flow
import { connect } from 'react-redux';
import {
  makeGetEnabledFulfillmentMethods,
} from '../../../core/config';
import {
  makeIsFulfillmentSelectorDisabled,
  makeGetUserLocation,
  makeGetProductLocation,
  makeGetUserLocationFulfillmentMethod,
} from '../../selectors';
import { storeFulfillmentMethod } from '../../action-creators';
import { type OwnProps, type StateProps, type DispatchProps } from './FulfillmentSelector.types';

/**
 * @param {Object} state The current application state.
 * @param {Object} props The component props.
 * @return {Object} The extended component props.
 */
function makeMapStateToProps() {
  const getUserLocation = makeGetUserLocation();
  const isFulfillmentSelectorDisabled = makeIsFulfillmentSelectorDisabled();
  const getProductLocation = makeGetProductLocation(true);
  const getUserLocationFulfillmentMethod = makeGetUserLocationFulfillmentMethod();
  const getEnabledFulfillmentMethods = makeGetEnabledFulfillmentMethods();

  /**
   * @param {Object} state The application state.
   * @param {Object} props The component props.
   * @returns {Object}
   */
  return (state, props) => ({
    shopFulfillmentMethods: getEnabledFulfillmentMethods(state),
    userFulfillmentMethod: getUserLocationFulfillmentMethod(state, props),
    location: getProductLocation(state, props) || getUserLocation(state),
    disabled: isFulfillmentSelectorDisabled(state, props),
  });
}

const mapDispatchToProps = {
  storeFulfillmentMethod,
};

export default connect<StateProps, DispatchProps, OwnProps>(
  makeMapStateToProps,
  mapDispatchToProps
);
