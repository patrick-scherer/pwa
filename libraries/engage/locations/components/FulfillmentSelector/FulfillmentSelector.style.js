import { css } from 'glamor';
import classNames from 'classnames';
import radioGroupStyles from '@shopgate/pwa-ui-shared/Form/RadioGroup/components/Item/style';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { variables: { gap: { big: bigGap = 16 } = {} } = {} } = themeConfig;

export const container = css({
  padding: `${(bigGap * 0.8).toFixed()}px ${bigGap}px ${bigGap}px`,
});

export const title = css({
  fontSize: '1rem',
  fontWeight: 500,
  marginBottom: 8,
});

const radioBase = css({
  // Removes the vertical padding, which are applied by default around radio groups
  paddingTop: 0,
  paddingBottom: 0,
  transition: 'opacity 250ms cubic-bezier(0.25, 0.1, 0.25, 1)',
});

export const radioGroup = radioBase.toString();

export const radioGroupDisabled = css(radioBase, {
  opacity: 0.25,
  pointerEvents: 'none',
}).toString();

export const radioItemLabelSpacer = css({
  whiteSpace: 'pre',
}).toString();

export const radioGroupItemLabel = classNames(radioGroupStyles.label, css({
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'space-between',
  alignItems: 'baseline',
}).toString());

export const radioItemLabel = css({
  fontSize: '0.875rem',
}).toString();

export const pickUpGroupContainer = css({
  display: 'flex',
  flexDirection: 'column',
  textAlign: 'right',
});
