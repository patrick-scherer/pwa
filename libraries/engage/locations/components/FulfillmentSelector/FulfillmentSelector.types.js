// @flow
import { type ProductId } from '../../../product';
import { DIRECT_SHIP, IN_STORE_PICKUP } from '../../constants';
import { type Location, type UserLocationFulfillmentMethod } from '../../locations.types';

export type Selection = typeof DIRECT_SHIP | typeof IN_STORE_PICKUP;

export type FulfillmentSelectorContextProps = {
  selection: Selection,
  selectedLocation: Location | null,
  location: Location | null,
  disabled: boolean,
  productId: ProductId,
  handleChange: (element: Selection, changeOnly: boolean) => Promise<void>,
  conditioner: any,
  fulfillmentPaths: string[],
  userFulfillmentMethod: string | null,
}

export type OwnProps = {
  conditioner: any,
  productId: string,
}

export type StateProps = {
  shopFulfillmentMethods?: string[] | null,
  disabled: boolean,
  location: Location | null,
  userFulfillmentMethod: string | null,
  fulfillmentPaths: string[],
}

export type DispatchProps = {
  storeFulfillmentMethod: (method: UserLocationFulfillmentMethod) => void,
}
