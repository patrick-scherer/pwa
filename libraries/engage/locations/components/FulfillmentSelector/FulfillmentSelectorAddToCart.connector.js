import { connect } from 'react-redux';
import { makeGetFulfillmentPaths } from '@shopgate/engage/core/config';
import { makeGetUserLocation } from '../../selectors';
import { PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP } from '../../constants';

/**
 * @return {Function}
 */
function makeMapStateToProps() {
  const getFulfillmentPaths = makeGetFulfillmentPaths();
  const getUserLocation = makeGetUserLocation();

  /**
   * @param {Object} state The application state.
   * @returns {Object}
   */
  return (state) => {
    const {
      fulfillmentMethod = PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP,
    } = getUserLocation(state) || {};

    return {
      fulfillmentPaths: getFulfillmentPaths(state),
      fulfillmentMethod,
    };
  };
}

export default connect(makeMapStateToProps);
