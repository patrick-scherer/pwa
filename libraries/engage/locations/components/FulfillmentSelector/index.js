// @flow
export { default as FulfillmentSelector } from './FulfillmentSelector';
