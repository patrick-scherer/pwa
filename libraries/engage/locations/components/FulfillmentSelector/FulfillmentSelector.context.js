// @flow
import { createContext } from 'react';
import { type FulfillmentSelectorContextProps } from './FulfillmentSelector.types';
import { DIRECT_SHIP } from '../../constants';

export const FulfillmentSelectorContext = createContext<FulfillmentSelectorContextProps>({
  selection: DIRECT_SHIP,
  selectedLocation: null,
  location: null,
  disabled: true,
  productId: null,
  handleChange: async () => { },
  conditioner: {},
  fulfillmentPaths: [],
  userFulfillmentMethod: null,
});
