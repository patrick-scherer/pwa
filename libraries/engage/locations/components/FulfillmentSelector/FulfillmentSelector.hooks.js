// @flow
import { useContext } from 'react';
import { type FulfillmentSelectorContextProps } from './FulfillmentSelector.types';
import { FulfillmentSelectorContext } from './FulfillmentSelector.context';

/**
 * @returns {Object} The fulfillment selector context.
 */
export function useFulfillmentSelectorState(): FulfillmentSelectorContextProps {
  return useContext(FulfillmentSelectorContext);
}
