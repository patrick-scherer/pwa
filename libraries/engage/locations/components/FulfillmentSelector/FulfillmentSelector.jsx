// @flow
import { hot } from 'react-hot-loader/root';
import React, { useCallback, useState } from 'react';
import { I18n, SurroundPortals, RadioGroup } from '@shopgate/engage/components';
import { Availability } from '@shopgate/engage/product';
import { FulfillmentSheet } from '../FulfillmentSheet';
import { StockInfo } from '../StockInfo';
import {
  PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP,
  PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP,
  PRODUCT_FULFILLMENT_SELECTOR,
} from '../../constants';
import FulfillmentSelectorItem from './FulfillmentSelectorItem';
import FulfillmentSelectorButtonChangeLocation from './FulfillmentSelectorButtonChangeLocation';
import FulfillmentSelectorAddToCart from './FulfillmentSelectorAddToCart';
import connect from './FulfillmentSelector.connector';
import * as styles from './FulfillmentSelector.style';
import { type OwnProps, type StateProps, type DispatchProps } from './FulfillmentSelector.types';

type Props = OwnProps & StateProps & DispatchProps;

/**
 * Renders a fulfillment selector box for fulfillment methods direct ship and pick up in store,
 * when fulfillment methods are set up for the product and pick up in store is one of them.
 * @param {Object} props The component props.
 * @param {string} props.productCode Code to the product or a variant.
 * @param {string[]} props.fulfillmentMethods All fulfillment methods provided for the product.
 * @param {Object} props.location Last location that was selected for the previous product/variant.
 * @returns {JSX}
 */
export const FulfillmentSelector = (props: Props) => {
  const {
    productId: productCode,
    fulfillmentMethods,
    location,
    conditioner,
    disabled,
    storeFulfillmentMethod,
  } = props;
  const directShip = 'product.fulfillment_selector.direct_ship';
  const pickUp = 'product.fulfillment_selector.pick_up_in_store';

  const [selection, setSelection] = useState(location.productInventory ? pickUp : directShip);
  const [selectedLocation, setSelectedLocation] = useState(null);
  const [isOpen, setIsOpen] = useState(false);

  // Update selected location for sheet close
  const handleClose = useCallback((newLocationData) => {
    setIsOpen(false);
    if (!newLocationData) {
      // Reset the UI back to directShip if there was no location selected already
      if (selectedLocation === null) {
        setSelection(directShip);
      }
    } else if (newLocationData.productCode === productCode) {
      // Update the selected location only when the selection was done for the same product.
      setSelectedLocation(newLocationData.location);
    }
  }, [selectedLocation, productCode]);

  /**
   * Whenever the pick-up selection is made, open the
   * store selector sheet and use the new location.
   */
  const handleChange = useCallback((elementName, changeOnly = false) => {
    // Run only external conditions
    conditioner.without('fulfillment-inventory').check().then((passed) => {
      if (!passed) {
        return;
      }

      let method = PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP;

      if (elementName === pickUp) {
        method = PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP;
      }

      setSelection(elementName);
      storeFulfillmentMethod(method);

      if (isOpen) {
        return;
      }

      if (elementName === directShip) {
        setSelectedLocation(null);
        return;
      }

      if (elementName === pickUp && location.code) {
        return;
      }

      setIsOpen(true);
      FulfillmentSheet.open({
        callback: handleClose,
        changeOnly,
      });
    });
  }, [conditioner, storeFulfillmentMethod, isOpen, location, handleClose]);

  if (!fulfillmentMethods) {
    return null;
  }

  return (
    <SurroundPortals portalName={PRODUCT_FULFILLMENT_SELECTOR} portalProps={{ productCode }}>
      <div
        className={styles.container}
        data-test-id="product-fulfillment-selector"
      >
        <div role="heading" aria-hidden className={styles.title}>
          <I18n.Text string="product.fulfillment_selector.heading" />
        </div>
        <RadioGroup
          name="product.fulfillment_selector"
          value={selection}
          className={disabled ? styles.radioGroupDisabled : styles.radioGroup}
          onChange={handleChange}
          isControlled
          direction="column"
        >
          <FulfillmentSelectorItem name={directShip}>
            <Availability
              productId={productCode}
              fulfillmentSelection={PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP}
            />
          </FulfillmentSelectorItem>
          <FulfillmentSelectorItem name={pickUp}>
            {location && (
              <div className={styles.pickUpGroupContainer}>
                <StockInfo location={selectedLocation || location} />
                {(selection === pickUp) && (
                  <FulfillmentSelectorButtonChangeLocation
                    onClick={() => handleChange(PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP, true)}
                  />
                )}
              </div>
            )}
          </FulfillmentSelectorItem>
        </RadioGroup>
      </div>
      <FulfillmentSelectorAddToCart
        conditioner={conditioner}
        location={selectedLocation || location}
      />
    </SurroundPortals>
  );
};

FulfillmentSelector.defaultProps = {
  disabled: true,
  fulfillmentMethods: null,
  location: null,
  storeFulfillmentMethod() { },
};

export default hot(connect(FulfillmentSelector));
