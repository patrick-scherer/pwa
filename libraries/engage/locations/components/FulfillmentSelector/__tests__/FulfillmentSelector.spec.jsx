import React from 'react';
import { shallow } from 'enzyme';
import {
  PRODUCT_FULFILLMENT_METHOD_ROPIS,
} from '../../../constants';
import { FulfillmentSheet } from '../../FulfillmentSheet';
import { FulfillmentSelector } from '../FulfillmentSelector';

jest.mock('../../FulfillmentSheet', () => ({
  FulfillmentSheet: {
    open: jest.fn(),
  },
}));
jest.mock('../FulfillmentSelectorAddToCart', () => 'FulfillmentSelectorAddToCart');
jest.mock('../FulfillmentSelectorItemDirectShip', () => 'FulfillmentSelectorItemDirectShip');
jest.mock('../FulfillmentSelectorItemReserve', () => 'FulfillmentSelectorItemReserve');

const conditioner = {
  check: () => Promise.resolve(true),
  without: () => conditioner,
};

describe('<FulfillmentSelector />', () => {
  it('should not render if the fulfillmentMethods do not contain inStorePickup', () => {
    const wrapper = shallow((
      <FulfillmentSelector
        productId="abc-123"
        location={{
          name: 'Test Store',
        }}
        conditioner={conditioner}
        shopFulfillmentMethods={[]}
        disabled={false}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.instance()).toEqual(null);
  });

  it('should render the fulfillment selector with two radio elements', () => {
    const wrapper = shallow((
      <FulfillmentSelector
        productId="abc-123"
        location={{
          name: 'Test Store',
        }}
        conditioner={conditioner}
        shopFulfillmentMethods={[
          PRODUCT_FULFILLMENT_METHOD_ROPIS,
        ]}
        fulfillmentPaths={[]}
        disabled={false}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('FulfillmentSelectorItem').length).toBe(2);
    expect(wrapper.find('[name="product.fulfillment_selector.direct_ship"]').length).toBe(1);
    expect(wrapper.find('[name="product.fulfillment_selector.pick_up_in_store"]').length).toBe(1);
  });

  it('should open the store selector sheet on selecting pick_up_in_store', async () => {
    const wrapper = shallow((
      <FulfillmentSelector
        productId="abc-123"
        location={{}}
        conditioner={conditioner}
        shopFulfillmentMethods={[
          PRODUCT_FULFILLMENT_METHOD_ROPIS,
        ]}
        disabled={false}
      />
    ));

    wrapper
      .find('[name="product.fulfillment_selector"]')
      .simulate('change', 'product.fulfillment_selector.pick_up_in_store');

    await new Promise(resolve => setImmediate(resolve));

    expect(FulfillmentSheet.open).toHaveBeenCalledTimes(1);
  });
});
