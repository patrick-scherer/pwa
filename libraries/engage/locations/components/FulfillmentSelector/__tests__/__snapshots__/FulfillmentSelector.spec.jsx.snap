// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FulfillmentSelector /> should not render if the fulfillmentMethods do not contain inStorePickup 1`] = `
<SurroundPortals
  portalName="product.fulfillment-selector"
  portalProps={
    Object {
      "productCode": "abc-123",
    }
  }
>
  <div
    className={
      Object {
        "data-css-12ywm8s": "",
      }
    }
    data-test-id="product-fulfillment-selector"
  >
    <div
      aria-hidden={true}
      className={
        Object {
          "data-css-1femajq": "",
        }
      }
      role="heading"
    >
      <Translate
        className={null}
        params={Object {}}
        role={null}
        string="product.fulfillment_selector.heading"
        transform={null}
      />
    </div>
    <RadioGroup
      className="css-11ghxnx"
      direction="column"
      errorText=""
      isControlled={true}
      label=""
      name="product.fulfillment_selector"
      onChange={[Function]}
      translateErrorText={true}
      value="product.fulfillment_selector.direct_ship"
    >
      <FulfillmentSelectorItem
        name="product.fulfillment_selector.direct_ship"
      >
        <FulfillmentSelectorItemDirectShip
          productId="abc-123"
          selected={true}
        >
          <Connect(Availability)
            fulfillmentSelection="directShip"
            productId="abc-123"
          />
        </FulfillmentSelectorItemDirectShip>
      </FulfillmentSelectorItem>
      <FulfillmentSelectorItem
        name="product.fulfillment_selector.pick_up_in_store"
      >
        <FulfillmentSelectorItemReserve
          location={
            Object {
              "name": "Test Store",
            }
          }
          selected={false}
        >
          <div
            className={
              Object {
                "data-css-15mvxgc": "",
              }
            }
          >
            <StockInfo
              className=""
              location={
                Object {
                  "name": "Test Store",
                }
              }
              showStoreName={true}
            />
          </div>
        </FulfillmentSelectorItemReserve>
      </FulfillmentSelectorItem>
    </RadioGroup>
  </div>
  <FulfillmentSelectorAddToCart
    conditioner={
      Object {
        "check": [Function],
        "without": [Function],
      }
    }
    location={
      Object {
        "name": "Test Store",
      }
    }
  />
</SurroundPortals>
`;

exports[`<FulfillmentSelector /> should render the fulfillment selector with two radio elements 1`] = `
<SurroundPortals
  portalName="product.fulfillment-selector"
  portalProps={
    Object {
      "productCode": "abc-123",
    }
  }
>
  <div
    className={
      Object {
        "data-css-12ywm8s": "",
      }
    }
    data-test-id="product-fulfillment-selector"
  >
    <div
      aria-hidden={true}
      className={
        Object {
          "data-css-1femajq": "",
        }
      }
      role="heading"
    >
      <Translate
        className={null}
        params={Object {}}
        role={null}
        string="product.fulfillment_selector.heading"
        transform={null}
      />
    </div>
    <RadioGroup
      className="css-11ghxnx"
      direction="column"
      errorText=""
      isControlled={true}
      label=""
      name="product.fulfillment_selector"
      onChange={[Function]}
      translateErrorText={true}
      value="product.fulfillment_selector.direct_ship"
    >
      <FulfillmentSelectorItem
        name="product.fulfillment_selector.direct_ship"
      >
        <FulfillmentSelectorItemDirectShip
          productId="abc-123"
          selected={true}
        >
          <Connect(Availability)
            fulfillmentSelection="directShip"
            productId="abc-123"
          />
        </FulfillmentSelectorItemDirectShip>
      </FulfillmentSelectorItem>
      <FulfillmentSelectorItem
        name="product.fulfillment_selector.pick_up_in_store"
      >
        <FulfillmentSelectorItemReserve
          location={
            Object {
              "name": "Test Store",
            }
          }
          selected={false}
        >
          <div
            className={
              Object {
                "data-css-15mvxgc": "",
              }
            }
          >
            <StockInfo
              className=""
              location={
                Object {
                  "name": "Test Store",
                }
              }
              showStoreName={true}
            />
          </div>
        </FulfillmentSelectorItemReserve>
      </FulfillmentSelectorItem>
    </RadioGroup>
  </div>
  <FulfillmentSelectorAddToCart
    conditioner={
      Object {
        "check": [Function],
        "without": [Function],
      }
    }
    location={
      Object {
        "name": "Test Store",
      }
    }
  />
</SurroundPortals>
`;
